corehq/ex-submodules/pillowtop/processors/elastic.py

```diff
@@ -78,16 +79,17 @@ def _datadog_timing(self, step):
         ], timing_buckets=(.03, .1, .3, 1, 3, 10))
 
 
-def send_to_elasticsearch(index, doc_type, doc_id, es_getter, name, data=None, retries=MAX_RETRIES,
-                          except_on_failure=False, update=False, delete=False, es_merge_update=False):
+def send_to_elasticsearch(index, doc_type, doc_id, es_getter, name, data=None,
+                          retries=MAX_RETRIES, propagate_failure=settings.UNIT_TESTING,
+                          update=False, delete=False, es_merge_update=False):
     """
     More fault tolerant es.put method
     kwargs:
         es_merge_update: Set this to True to use Elasticsearch.update instead of Elasticsearch.index
             which merges existing ES doc and current update. If this is set to False, the doc will be replaced
 
     """
-    data = data if data is not None else {}
+    data = {k: v for k, v in data.items() if k != '_id'} if data is not None else {}
```

Comment:
how confident are you that this will work/we aren't using `_id`? I see that we do specify `doc_id`, but I'm not sure if ES sets that automatically (its pulled from teh couch doc right?)

Your commit says that this is "mandated" by ES 2.x but not why its mandated. Can you explain that/link to documentation?

Reply:
Yeah, sorry. I could not find this  explicitly in the breaking changes notes, but in 2.x (but not 1.x) having an `_id` field in the body results in the following error (taken from ES logs):

> MapperParsingException[Field [_id] is a metadata field and cannot be added inside a document. Use the index API request parameters.]

Note that in such requests, the _id is already specified in the url or some other part of the request; it's just not allowed to also show up in the doc "body" part of the request.

Reply:
Ah cool. So it sounds like `_id` is something that both CouchDB and ES happen to use as the definition of primary key for their documents. I had gone quite a while thinking that was just a CouchDB thing that we copied over to ES